### ISSUE TITLE
PERSON: Bis Datum auf Rolle verwirrend

### ISSUE THREAD

**Issue by nchiapol** [hitobito/hitobito#1743]
Vielen Dank für #1714 und die Möglichkeit das End-Datum zu setzen. 

Mindestens bei uns (Cevi) ist damit das User-Interface nun aber verwirrend. Ich habe nun zwei Möglichkeiten eine Rolle zu beenden:
  1. Ich klicke auf editieren und setze ein Enddatum in der Vergangenheit. Die Rolle wird dann automatisch gelöscht und das korrekte Enddatum im Verlauf angezeigt.
  2. Ich klicke auf den Löschen-Knopf. Die Rolle wird dann automatisch gelöscht und das aktuelle Datum als Enddatum angezeigt. (was wohl meist nicht stimmt.) Da ich im Moment keine Möglichkeit habe die Daten im Verlauf zu ändern, ist das deshalb wohl meist das falsche vorgehen.

Ich sehe verschiedene mögliche Verbesserungen:
  - Der Löschen-Knopf öffnet einen Dialog, in dem ich zwingend ein Enddatum eintragen muss. (default: heute)
  - Der Löschen-Knopf wird einfach ganz deaktiviert, Rollen können nur noch über Bearbeiten und Eintragen des Enddatums gelöscht werden.
  - Es wird ermöglicht die Daten im Verlauf auch noch nachträglich zu ändern (wohl unabhängig der obigen Vorschläge sinnvoll.) Ich würde diese Möglichkeit an "layer_full" knüpfen.

**Comment by Michael-Schaer** [comment]
Merci fürs Ansprechen dieser Änderung! 

> * Der Löschen-Knopf wird einfach ganz deaktiviert, Rollen können nur noch über Bearbeiten und Eintragen des Enddatums gelöscht werden.

Der Rolle-Löschen-Knopf finde ich Usability-Mässig recht gut gelungen. Wenn jemand aus der Gruppe ausgestiegen ist, habe ich direkt eine Möglichkeit, dies zu erledigen (ohne Zusatzinformationen angeben zu müssen). 

Der Button fördert aus meiner Sicht auch das korrekte Verhalten. Nämlich die Rolle zu entfernen und nicht z.B. die Felder der Person zu überschreiben oder den Löschen-Knopf auf der Person zu drücken.

**Comment by nchiapol** [comment]
@Michael-Schaer: Ich finde meinen zweiten Vorschlag auch nicht überzeugend - aber besser als der Status quo. Und der lässt sich wahrscheinlich mit wenig Aufwand umgehend umsetzen.  Die erste Variante ist aber klar die korrekte Lösung.

**Comment by betsim** [comment]
Aus Sicht SBV gäbe es bei der Rollen-Administration tatsächlich auch noch Verbesserungspotential. Generell würde ich das Rollen-Handling in den Tabs INFO und VERLAUF vereinheitlichen. Die unterschiedliche Handhabung verwirrt viele Benutzer. Neben aktiven Rollen zu bearbeiten, soll es genauso einfach möglich sein, vergangene Rollen bearbeiten zu können. Heute müssen falsch erfasste Rollen gelöscht und nochmals neu erfasst werden.

Ich würde die beiden Funktionen "bearbeiten" und "löschen" beibehalten - aber in beiden Tabs. Ich finde den Vorschlag gut, den Bestätigungs-Dialog beim Löschen dahingehend auszubauen, dass gefragt wird, ob die Rolle per xxx (Angabe des heutigen Datums vorausgefüllt) entfernt werden soll. Es bleibt für den User der Klick auf OK, wie bisher - oder er ändert das Datum gleich hier.

> Übrigens: Was jetzt neu möglich ist, sind Löschdaten in der Zukunft. (Das müssten wir wohl noch im Changelog ergänzen)

Super, sehr gut! Aber ist es wirklich sinnvoll, wenn die Rolle - obwohl das Datum in der Zukunft liegt - und damit die entsprechenden Rechte trotzdem per sofort entzogen wird?

Weiter stellt sich mir die Frage, ob wirklich so viele Buttons "+ Rolle hinzufügen" im Tab INFO nötig sind: 
![Aktive Rollen - jetzt](https://user-images.githubusercontent.com/1453303/169784345-2662aa58-1fc6-4843-a435-495bda1ed21c.png)
...oder ob ein einziger Button unterhalb "Aktive Rollen" nicht genügen würde. 

Ausserdem besteht beim SBV offenbar noch eine Besonderheit (?oder haben das die anderen auch?), dass auf dem Tab VERLAUF direkt auch eine neue Rolle erfasst werden kann, wobei nur die Gruppe ausgewählt werden kann und immer die Rolle "Mitglied" erfasst wird:
![Neue Rolle - VERLAUF](https://user-images.githubusercontent.com/1453303/169785216-67e016dd-3bdc-4193-a9ea-5f7d877bbf86.png)
Das wurde ursprünglich so gemacht, um die Einträge aus den bestehenden Musikerpässen zu übernehmen. Ich sehe aber keinen Grund, weshalb dieser Dialog nicht mit dem Standard-Rolle-Hinzufügen-Dialog vom Tab INFO ersetzt werden kann:
![Neue Rolle - INFO](https://user-images.githubusercontent.com/1453303/169785328-08b2de91-dce4-4d1f-9745-e9b8e1f9e27c.png)

**Comment by carlobeltrame** [comment]
> > Übrigens: Was jetzt neu möglich ist, sind Löschdaten in der Zukunft. (Das müssten wir wohl noch im Changelog ergänzen)
> 
> Super, sehr gut! Aber ist es wirklich sinnvoll, wenn die Rolle - obwohl das Datum in der Zukunft liegt - und damit die entsprechenden Rechte trotzdem per sofort entzogen wird?

Ich habs bisher nicht selber getestet, aber dem Code zufolge müsste ich meine Rechte bis zum Löschdatum noch behalten. Also die Rechte werden nicht sofort entzogen, sondern erst am eingetragenen Datum. Bitte unbedingt auf der Integration testen und melden falls es sich anders verhält als erwartet. Merci fürs Interesse!

**Comment by betsim** [comment]
Habs nochmals getestet und du hast Recht Carlo, die Rolle und die Rechte werde nicht sofort entzogen - da hab ich irgendwas falsch gesehen heute Morgen, sorry.

**Comment by Michael-Schaer** [comment]
> Wäre es wirklich schlimm wenn wir diesen Dialog ersetzen würden durch einen neuen Dialog der noch das Setzen des genauen Enddatums, falls nicht der heutige Tag, erlaubt?

Nein, gegen diese Variante habe ich nichts. Einfach den Rolle-Löschen Button zu entfernen, fände ich usability mässig sehr ungünstig.
 
> Übrigens: Was jetzt neu möglich ist, sind Löschdaten in der Zukunft. (Das müssten wir wohl noch im Changelog ergänzen)

Wurde eigentlich der Performance-Inpact dieser Funktion getestet? Das Feature wurde doch schon mal abgelehnt, weil dann grössere Exporte zu lange dauerten.

**Comment by carlobeltrame** [comment]
> > Übrigens: Was jetzt neu möglich ist, sind Löschdaten in der Zukunft. (Das müssten wir wohl noch im Changelog ergänzen)
> 
> Wurde eigentlich der Performance-Inpact dieser Funktion getestet? Das Feature wurde doch schon mal abgelehnt, weil dann grössere Exporte zu lange dauerten.

Ich wüsste nichts davon, dass wir die Performance ausführlich getestet hätten (@TheWalkingLeek ?). Ich kann mich aber auch nicht daran erinnern dass wir es aus diesem Grund abgelehnt hatten, hast du das noch irgendwo?
Wir mussten für die Implementation eine eingesetzte Library leicht modifizieren, weil diese Library das nicht umsetzen will (https://github.com/rubysherpas/paranoia/issues/480). Möglicherweise wurden als Begründung euch gegenüber mal Performance-Bedenken genannt, weil in der Datenbank ein Vergleich mit dem heutigen Datum vordergründig aufwändiger ist als ein Vergleich mit NULL. Aber die Library unterstützt auch andere [Modi](https://github.com/ActsAsParanoid/acts_as_paranoid/blob/master/lib/acts_as_paranoid/core.rb#L48..L49), die in etwas komplexeren SQL-Queries resultieren. Also wir müssen es vermutlich einfach ausprobieren.

Gerade ist mir aufgefallen dass mit diesem Feature vermutlich noch nicht alles stimmt, bzgl. Code der direkt nach dem Löschen der Rolle laufen sollte. Ich eröffne dazu noch ein Issue.

**Comment by nchiapol** [comment]
scheint mir nach wie vor verbesserungswürdig
